### ISSUE TITLE
In some cases, clippy's `manual_map` lint suggests changes that don't compile

### ISSUE THREAD

**Issue by dcormier** [rust-lang/rust-clippy#7413]
I tried this code:

```rust
use std::collections::HashMap;

fn main() {
    #[derive(Debug)]
    struct Thing {
        name: Option<&'static str>,
    }

    let map: HashMap<_, _> = vec![
        Thing { name: Some("a") },
        Thing { name: None },
        Thing { name: Some("b") },
    ]
    .into_iter()
    // Clippy suggests changing this manual map
    // https://rust-lang.github.io/rust-clippy/master/index.html#manual_map
    .filter_map(|thing| match &thing.name {
        Some(name) => Some((name.to_string(), thing)),
        None => None,
    })
    // Clippy's recommendation doesn't compile:
    // .filter_map(|thing| thing.name.as_ref().map(|name| (name.clone(), thing)))
    .collect();

    println!("{:?}", map);
}
```

[Playground.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1b3ecd4156b5d4432193bcb16fb64f13)

I expected to see this happen: Clippy would have no complaints.

Instead, this happened: Clippy complained about the [manual map](https://rust-lang.github.io/rust-clippy/master/index.html#manual_map):

```
warning: manual implementation of `Option::map`
  --> src/main.rs:17:25
   |
17 |       .filter_map(|thing| match &thing.name {
   |  _________________________^
18 | |         Some(name) => Some((name.to_string(), thing)),
19 | |         None => None,
20 | |     })
   | |_____^ help: try this: `thing.name.as_ref().map(|name| (name.to_string(), thing))`
   |
   = note: `#[warn(clippy::manual_map)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_map
```

But the suggested change doesn't compile.

```
error[E0505]: cannot move out of `thing` because it is borrowed
  --> src/main.rs:22:49
   |
22 |     .filter_map(|thing| thing.name.as_ref().map(|name| (name.clone(), thing)))
   |                         ----------          --- ^^^^^^                ----- move occurs due to use in closure
   |                         |                   |   |
   |                         |                   |   move out of `thing` occurs here
   |                         |                   borrow later used by call
   |                         borrow of `thing.name` occurs here
```

### Meta

- `cargo clippy -V`: clippy 0.1.53 (53cb7b09 2021-06-17)
- `rustc -Vv`:
```
rustc 1.53.0 (53cb7b09b 2021-06-17)
binary: rustc
commit-hash: 53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b
commit-date: 2021-06-17
host: x86_64-apple-darwin
release: 1.53.0
LLVM version: 12.0.1
```
